### PR TITLE
Feature/dkr 1 hangman

### PR DESCRIPTION
**Pull request by skorodenko**

### CODE REVIEW

File: hangman.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+             1 - with hints
+             2 - without hints
+    """
+    question = input(
+                     "Enter 1, if you want to play hangman with hints\n"+
+                     "Enter 0, if you want to play hangman without hints\n"+
+                     ">>>> "
+                    )
+    if question not in [str(ReturnQuestion.hangman_without_hints.value), 
+                        str(ReturnQuestion.hangman_with_hints.value)]:
```
Comment (Hozhyi) [inline hangman.py:272]: I'd use here a set of raw strings, e.g. {'1', '0'} for input checking

Comment (Hozhyi) [reply hangman.py:272]: And than I'd convert it to appropriate Enum

File: hangman.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+import string
+from enum import Enum
+
+WORDLIST_FILENAME = "words.txt"
+
+
+class BaseVariables():
+    """
+    This class creates an object with base variables, requiered for function
+    to work properly
```
Comment (Hozhyi) [inline hangman.py:17]: Effectively - all classes are required for your program, so it has no sense to mention it in the description.

File: hangman.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+def game_start():
+    """
+    This function starts the game.
+    """
+    stop = "n"
+    while stop != "y":
+        secret_word = choose_word(wordlist)
+        print("Welcome to the game Hangman!")
+        print(f"I am thinking of a word that is {len(secret_word)} letters long.")
+        print("-----------")
```
Comment (Hozhyi) [inline hangman.py:287]: Here you have a greeting, in next function - you ask for some question. Do you think that current output and following input - are different enough to split it? I think it's one part of logic, so it should be placed near each other

File: hangman.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+        question = question_input()
+        variables = BaseVariables(secret_word, question)
+        stop = game_restart(variables)
+
+
+def hangman(variables):
```
Comment (Hozhyi) [inline hangman.py:293]: Effectively - now this function already has a different role, so it makes sense to rename it

File: hangman.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    * After each guess, you should display to the user the 
+      partially guessed word so far.
+    
+    Follows the other limitations detailed in the problem write-up.
+    """
+    print(f"You have {variables.guesses_remaining} guesses left.")
+    print("Available letters:",get_available_letters(variables.letters_guessed))
+    guess = input("Please guess a letter: ").strip()
+    # used strip in case that user accidently inputs space
+    if guess == "*" and variables.question == 1:
```
Comment (Hozhyi) [inline hangman.py:324]: It's hard to read that `variables.question == 1`. So we have some `question`, and it's value can be equal to `1`? Question - is a sentence, that at least ends with a question mark.

File: hangman.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+                                  get_guessed_word(variables.secret_word, 
+                                                   variables.letters_guessed))
+            print(get_guessed_word(variables.secret_word,
+                                   variables.letters_guessed))
+            return 
+        else:
+            print("You haven`t guessed a single letter. There is no sense in hint.")
+            return 
+    elif guess == "*":
+        print("You play hangman without hints.")
```
Comment (Hozhyi) [inline hangman.py:339]: I can remember that PO asked you to implement scenario where you remind to user this information only once, and next time it should be translated as a warning

File: hangman.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+    return ReturnValue.valid_guess
+    
+
+def guess_try(guess, variables):
+    """
+    This function checks whether the guess contains in secret word.
+    guess: string, input from user
+    variables: object of BaseVariables class
+    returns: text output depending on guess,
+             dictionary, variables
```
Comment (Hozhyi) [inline hangman.py:144]: From your description - it seems like you return a text and a dictionary. But according to your code - you return nothing

**Review by Hozhyi — CHANGES_REQUESTED** [review]

File: hangman.py (hunk 2 above)
Comment (skorodenko) [reply hangman.py:17]: Corrected.

File: hangman.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+        self.guesses_remaining = 6
+        self.warnings_remaining = 3
+        self.letters_guessed = set()
+        self.secret_word = secret_word
+        self.question = question
+        self.wordlist = wordlist
+
+
+class ReturnValue(Enum):
+    not_latin = 1
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:29]: Changed case to uppercase with undersocres.

File: hangman.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+    def __init__(self, secret_word, question):
+        self.guesses_remaining = 6
+        self.warnings_remaining = 3
+        self.letters_guessed = set()
+        self.secret_word = secret_word
+        self.question = question
+        self.wordlist = wordlist
+
+
+class ReturnValue(Enum):
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:28]: Class name changed to CheckValue, and added comment that this class used only as return value for function guess_check.

File: hangman.py [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+    returns: string (of letters), comprised of letters that represents which letters has not
+             yet been guessed.
+    """
+    def get_unique(x):
+        nonlocal letters_guessed
+        if x in letters_guessed:
+            return False
+        return True
+    # Removing guessed letters from alphabet 
+    available_letters = [letter for letter in filter(get_unique, string.ascii_lowercase)]
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:113]: Changed.

File: hangman.py [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+    This function returns word with underscore symbols.
+    secret_word: string, the word the user is guessing
+    letters_guessed: set of letters which has been guessed so far
+    returns: string, comprised of letters, underscores (_), and spaces that represents
+      which letters in secret_word have been guessed so far.
+    """
+    # getting set of all letters that were not guessed
+    check = set(secret_word).difference(letters_guessed) 
+    for letter in check:
+        secret_word = secret_word.replace(letter,"_ ")
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:96]: Removed replace method, created for cycle with if/else statements.

File: hangman.py (hunk 7 above)
Comment (skorodenko) [reply hangman.py:144]: Changed description, added variables.letters_guessed to return statement.

File: hangman.py [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+        print("No matches found.")
+        return
+
+
+def display(output, variables):
+    """
+    This function is created to display data with various print functions.
+    output: Enum class with in value, option to triger particular text output.
+    variables: object of BaseVariables class
+    returns: nothing, but changes variables object.
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:212]: Changed description, added variables object to return statement.

File: hangman.py [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+    print(
+         f"You have {variables.warnings_remaining} warnings left:\n"+
+         get_guessed_word(variables.secret_word,variables.letters_guessed)
+         )
+    return 
+
+        
+def game_restart(variables):
+    """
+    This function restarts the game.
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:238]: Separated this function into game_start and game_restart. 1st starts the game, 2nd restarts it.

File: hangman.py [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+    variables: object of BaseVariables class
+    returns: str, stop
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:240]: Changed description to returns: stop(string).

File: hangman.py [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+    returns: str, stop  
+    """
+    while variables.guesses_remaining > 0:
+        if is_word_guessed(variables.secret_word, variables.letters_guessed):
+            print(
+                  "Congratulations, you won!\n"+
+                  "Your total score for this game is: "+
+                  f"{variables.guesses_remaining*len(set(variables.secret_word))}"
+                 )
+            stop = input("Do you want to quit the game. y/n \n")
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:249]: Due to the separation of function game_restart into game_start and game_restart, this duplication removed.

File: hangman.py [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+            break
+        hangman(variables)
+        print("-----------")
+    else:
+        print(f"Sorry, you ran out of guesses. The word was {variables.secret_word}.")
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:254]: Created separated function function game_result, which displays one of the messages, depending on game result.

File: hangman.py [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+        stop = input("Do you want to quit the game. y/n \n")
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:255]: Solved.

File: hangman.py [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+    return stop
+
+
+def question_input():
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:259]: Resolved.

File: hangman.py [hunk 19]
```diff
@@ -0,0 +1,354 @@
...
+    """
+    This function accepts input from user (integer).
+    returns: Enum with int values, option for game.
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:262]: Changed return to Enum class.

File: hangman.py [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+             1 - with hints
+             2 - without hints
+    """
+    question = input(
+                     "Enter 1, if you want to play hangman with hints\n"+
+                     "Enter 0, if you want to play hangman without hints\n"+
+                     ">>>> "
+                    )
+    if question not in [str(ReturnQuestion.hangman_without_hints.value),
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:271]: Changed.

File: hangman.py (hunk 1 above)
Comment (skorodenko) [reply hangman.py:272]: Changed.

File: hangman.py [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+    """
+    question = input(
+                     "Enter 1, if you want to play hangman with hints\n"+
+                     "Enter 0, if you want to play hangman without hints\n"+
+                     ">>>> "
+                    )
+    if question not in [str(ReturnQuestion.hangman_without_hints.value), 
+                        str(ReturnQuestion.hangman_with_hints.value)]:
+        print("You should enter 1 or 0.")
+        question = question_input()
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:274]: If it is not a rude mistake, I would like to leave it as it is. Otherwise, let me know.

File: hangman.py (hunk 3 above)
Comment (skorodenko) [reply hangman.py:287]: Removed redundant splitter.

File: hangman.py (hunk 4 above)
Comment (skorodenko) [reply hangman.py:293]: Changed hangman to hangman_combined.

File: hangman.py [hunk 22]
```diff
@@ -0,0 +1,354 @@
...
+        print(f"I am thinking of a word that is {len(secret_word)} letters long.")
+        print("-----------")
+        question = question_input()
+        variables = BaseVariables(secret_word, question)
+        stop = game_restart(variables)
+
+
+def hangman(variables): 
+    """
+    Hangman combined with Hangman with hints.
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:295]: Updated description and removed condition of hangman task.

File: hangman.py (hunk 5 above)
Comment (skorodenko) [reply hangman.py:324]: BaseVariables class now creates and object with "option" value not "question".

File: hangman.py (hunk 6 above)
Comment (skorodenko) [reply hangman.py:339]: Resolved.

File: hangman.py [hunk 23]
```diff
@@ -0,0 +1,369 @@
...
+class BaseVariables():
+    """
+    This class creates an object with base variables.
+    """
+    def __init__(self, secret_word, question):
+        self.guesses_remaining = 6
+        self.warnings_remaining = 3
+        self.letters_guessed = set()
+        self.secret_word = secret_word
+        self.option = question.value
```
Comment (Hozhyi) [inline hangman.py:23]: In case when wrong input and then - write input, we have an exception here

File: hangman.py [hunk 24]
```diff
@@ -0,0 +1,369 @@
+# Problem Set 2, hangman.py
```
Comment (Hozhyi) [inline hangman.py:1]: This program doesn't do a penalty of 2 attempts on vowels

**Review by Hozhyi — CHANGES_REQUESTED** [review]

File: hangman.py [hunk 25]
```diff
@@ -0,0 +1,369 @@
...
+
+WORDLIST_FILENAME = "words.txt"
+
+
+class BaseVariables():
+    """
+    This class creates an object with base variables.
+    """
+    def __init__(self, secret_word, question):
+        self.guesses_remaining = 6
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:19]: Corrected.

File: hangman.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:23]: Renamed question and optio to type_of_game

Comment (skorodenko) [reply hangman.py:23]: Exception resolved.

File: hangman.py [hunk 26]
```diff
@@ -0,0 +1,369 @@
...
+        self.warnings_remaining = 3
+        self.letters_guessed = set()
+        self.secret_word = secret_word
+        self.option = question.value
+        self.wordlist = wordlist
+        self.attempt = 0
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:25]: attempt renamed to hint_attempt

File: hangman.py [hunk 27]
```diff
@@ -0,0 +1,369 @@
...
+    """
+    my_word: string with _ characters, current guess of secret word
+    other_word: string, regular English word
+    variables: object of BaseVariables class
+    returns: boolean, True if all the actual letters of my_word match the 
+        corresponding letters of other_word, or the letter is the special symbol
+        _ , and my_word and other_word are of the same length;
+        False otherwise: 
+    """
+    my_word = my_word.replace(" ", "")
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:179]: Repalced.

File: hangman.py [hunk 28]
```diff
@@ -0,0 +1,369 @@
...
+    if len(my_word) == len(other_word):
+        for my_word_i, other_word_i in zip(range(len(my_word)),other_word):
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:181]: Simplified.

File: hangman.py [hunk 29]
```diff
@@ -0,0 +1,369 @@
...
+            if my_word[my_word_i] == "_":
+                if other_word_i in variables.letters_guessed:
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:183]: Due to the simplification of algorithm, comments are not needed.

File: hangman.py [hunk 30]
```diff
@@ -0,0 +1,369 @@
...
+
+
+def display(output, variables):
+    """
+    This function is created to display data with various print functions.
+    output: Enum class with in value, option to triger particular text output.
+    variables: object of BaseVariables class
+    returns: variables object.
+    """
+    if output == CheckValue.NOT_LATIN.value:
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:224]: Reduced the usage of .value in all places, where it is possible.

File: hangman.py [hunk 31]
```diff
@@ -0,0 +1,369 @@
...
+        print("Oops! That is not a valid character.")
+    if output == CheckValue.LENGTH_RESTRICTION.value:
+        print("Oops! You can't guess more than 1 character at once.")
+    if output == CheckValue.GUESS_GUESSED.value:
+        print("Oops! You've already guessed that letter.")
+    if variables.warnings_remaining == 0:
+        variables.guesses_remaining -= 1
+        variables.warnings_remaining = 3
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:232]: Removed.

File: hangman.py [hunk 32]
```diff
@@ -0,0 +1,369 @@
...
+
+
+def game_restart(variables):
+    """
+    This function restarts the game.
+    variables: object of BaseVariables class
+    returns: stop(string)  
+    """
+    game_start(variables)
+    stop = input("Do you want to quit the game. y/n \n")
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:288]: Implemented re-ask system.

File: hangman.py (hunk 24 above)
Comment (skorodenko) [reply hangman.py:1]: Resolved.

File: hangman.py [hunk 33]
```diff
@@ -0,0 +1,405 @@
...
+        self.letters_guessed = set()
+        self.secret_word = secret_word
+        self.type_of_game = type_of_game
+        self.wordlist = wordlist
+        self.hint_attempts = INITIAL_HINT_ATTEMPTS
+        self.vowels = VOWELS_LIST
+        self.hint = HINT_SYMBOL
+        self.gap = GAP_SYMBOL
+        self.yes = YES_ANSWER
+        self.no = NO_ANSWER
```
Comment (Hozhyi) [inline hangman.py:38]: I don't think you need to have it here if you have it in the global scope. Also I'd recommend to have these variables not in object, but in class

File: hangman.py [hunk 34]
```diff
@@ -0,0 +1,405 @@
...
+
+WORDLIST_FILENAME = "words.txt"
+VOWELS_LIST = ["a","e","i","o","u"]
+INITIAL_GUESSES = 6
+ININITAL_WARNINGS = 3
+INITIAL_HINT_ATTEMPTS = 1
+GAP_SYMBOL = "_ "
+HINT_SYMBOL = "*"
+YES_ANSWER = "y"
+NO_ANSWER = "n"
```
Comment (Hozhyi) [inline hangman.py:19]: I'm not sure it's a good practice to store example of user input in constants...

File: hangman.py [hunk 35]
```diff
@@ -0,0 +1,405 @@
...
+import random
+import string
+from enum import Enum
+
+WORDLIST_FILENAME = "words.txt"
+VOWELS_LIST = ["a","e","i","o","u"]
+INITIAL_GUESSES = 6
+ININITAL_WARNINGS = 3
+INITIAL_HINT_ATTEMPTS = 1
+GAP_SYMBOL = "_ "
```
Comment (Hozhyi) [inline hangman.py:16]: I believe a space is not needed there. There should be only a gap symbol

File: hangman.py [hunk 36]
```diff
@@ -0,0 +1,405 @@
...
+# Collaborators:
+# Time spent: 4 days
+
+# Hangman Game
+import random
+import string
+from enum import Enum
+
+WORDLIST_FILENAME = "words.txt"
+VOWELS_LIST = ["a","e","i","o","u"]
```
Comment (Hozhyi) [inline hangman.py:12]: Do you think that list - is the best container here?

File: hangman.py [hunk 37]
```diff
@@ -0,0 +1,405 @@
...
+    variables: object of BaseVariables class
+    returns: text output depending on guess,
+             varables.letters_guessed
+    """
+    variables.letters_guessed.add(guess)
+    if guess in variables.secret_word:
+        print("Good guess:",
+              get_guessed_word(variables.secret_word,
+                               variables.letters_guessed,
+                               variables.gap             ))
```
Comment (Hozhyi) [inline hangman.py:175]: Why do you have so many spaces in the end? (it's not only here, I can see)

File: hangman.py [hunk 38]
```diff
@@ -0,0 +1,405 @@
...
+    else:
+        # if letter is vowel -2 guesses
+        if guess in variables.vowels:
+            variables.guesses_remaining -= 2
```
Comment (Hozhyi) [inline hangman.py:179]: I still can see that you change an input object in scope of your function. In all such cases - return that object (`return variables`)

File: hangman.py [hunk 39]
```diff
@@ -0,0 +1,405 @@
...
+
+def game_start(variables):
+    """
+    This function starts the game.
+    variables: object of BaseVariables class
+    returns: nothing 
+    """
+    while variables.guesses_remaining > 0:
+        if is_word_guessed(variables.secret_word, variables.letters_guessed):
+            game_result(True, variables)
```
Comment (Hozhyi) [inline hangman.py:275]: Try to avoid having game_result output in scope of game loop

File: hangman.py [hunk 40]
```diff
@@ -0,0 +1,405 @@
...
+    else:
+        print(f"Sorry, you ran out of guesses. The word was {variables.secret_word}.")
+
+
+def game_restart(variables):
+    """
+    This function restarts the game.
+    variables: object of BaseVariables class
+    returns: stop(string)  
+    """
```
Comment (Hozhyi) [inline hangman.py:307]: I see that here you have implemented 2 processes:
1. you call a game_start
2. You ask user whether to continue
Try to have only 1 process in that function

**Review by Hozhyi — CHANGES_REQUESTED** [review]

File: hangman.py (hunk 36 above)
Comment (skorodenko) [reply hangman.py:12]: Changed to set.

File: hangman.py (hunk 35 above)
Comment (skorodenko) [reply hangman.py:16]: This space is needed, if you want to have your unguessed word look like ******, you set this constant to "*", if you want to have such word * * * * * * then you the constant to " * ".

File: hangman.py (hunk 34 above)
Comment (skorodenko) [reply hangman.py:19]: It is easier to modify such things without getting inside of code. Someone might like to have yes/no answer, another person would like to have y/n. Storing such constant will be a great solution for both types.

File: hangman.py (hunk 33 above)
Comment (skorodenko) [reply hangman.py:38]: Moved.

File: hangman.py (hunk 37 above)
Comment (skorodenko) [reply hangman.py:175]: Removed.

File: hangman.py (hunk 38 above)
Comment (skorodenko) [reply hangman.py:179]: Changed.

File: hangman.py (hunk 39 above)
Comment (skorodenko) [reply hangman.py]: Corrected.

File: hangman.py (hunk 40 above)
Comment (skorodenko) [reply hangman.py:301]: Corrected.

File: hangman.py [hunk 41]
```diff
@@ -0,0 +1,400 @@
...
+        type_of_game = input(
+                         "Enter 1, if you want to play hangman with hints\n"+
+                         "Enter 0, if you want to play hangman without hints\n"+
+                         ">>>> "
+                            )
+        if type_of_game not in set(("1", "0")):
+            print("You should enter 1 or 0.")
+            continue
+        else:
+            if int(type_of_game) == TypeOfGame.HANGMAN_WITH_HINTS.value:
```
Comment (Hozhyi) [inline hangman.py:330]: I believe it's will be easier to read the following: `if type_of_game == '1'`

File: hangman.py [hunk 42]
```diff
@@ -0,0 +1,400 @@
...
+
+def type_of_game_input():
+    """
+    This function accepts input from user (integer).
+    returns: Enum with int values, option for game.
+             1 - with hints
+             0 - without hints
+    """
+    valid = False
+    while valid != True:
```
Comment (Hozhyi) [inline hangman.py:320]: According to PEP it is not recommended to compare with True or False

File: hangman.py [hunk 43]
```diff
@@ -0,0 +1,400 @@
...
+            if int(type_of_game) == TypeOfGame.HANGMAN_WITH_HINTS.value:
+                type_of_game = TypeOfGame.HANGMAN_WITH_HINTS
+                valid = True
+            else:
+                type_of_game = TypeOfGame.HANGMAN_WITHOUT_HINTS
+                valid = True 
+    return type_of_game
+
+
+def game(YES_ANSWER, NO_ANSWER):
```
Comment (Hozhyi) [inline hangman.py:339]: Why the input parameter is named like a constant? Input parameters have different naming convention then constants

**Review by Hozhyi — CHANGES_REQUESTED** [review]

File: hangman.py [hunk 44]
```diff
@@ -0,0 +1,400 @@
...
+# Collaborators:
+# Time spent: 4 days
+
+# Hangman Game
+import random
+import string
+from enum import Enum
+
+WORDLIST_FILENAME = "words.txt"
+VOWELS_LIST = set(("a","e","i","o","u"))
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:12]: Changed.

File: hangman.py (hunk 34 above)
Comment (skorodenko) [reply hangman.py:19]: Removed yes/no constants.

File: hangman.py (hunk 35 above)
Comment (skorodenko) [reply hangman.py:16]: Separated GAP_SYMBOL -> GAP_SYMBOL and GAP_SPACE constants.

File: hangman.py [hunk 45]
```diff
@@ -0,0 +1,400 @@
...
+        self.guesses_remaining = INITIAL_GUESSES
+        self.warnings_remaining = ININITAL_WARNINGS
+        self.letters_guessed = set()
+        self.secret_word = secret_word
+        self.type_of_game = type_of_game
+        self.wordlist = wordlist
+        self.hint_attempts = INITIAL_HINT_ATTEMPTS
+
+
+class CheckValue(Enum):
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:42]: Changed.

File: hangman.py [hunk 46]
```diff
@@ -0,0 +1,400 @@
...
+    """
+    Enum class for return values in function guess_check.
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:44]: Changed.

File: hangman.py [hunk 47]
```diff
@@ -0,0 +1,400 @@
...
+def guess_check(guess, letters_guessed):
+    """
+    This function checks whether the guess is valid.
+    guess: string, input from user
+    letters_guessed: set of letters which has been guessed so far
+    returns: Enum class with int value; 
+        1 if guess is not latin letter;
+        2 if guess length is > 1;
+        3 if guess has already been guessed;
+        0 if guess is valid.
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:152]: Removed.

File: hangman.py [hunk 48]
```diff
@@ -0,0 +1,400 @@
...
+    other_word: string, regular English word
+    variables: object of BaseVariables class
+    returns: boolean, True if all the actual letters of my_word match the 
+        corresponding letters of other_word, or the letter is the special symbol
+        _ , and my_word and other_word are of the same length;
+        False otherwise: 
+    """
+    if len(my_word) == len(other_word):
+        for i in range(len(my_word)):
+            if my_word[i] == BaseVariables.gap.strip():
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:202]: Removed.

File: hangman.py (hunk 42 above)
Comment (skorodenko) [reply hangman.py:320]: valid != True --> not valid.

File: hangman.py (hunk 41 above)
Comment (skorodenko) [reply hangman.py:330]: Changed.

File: hangman.py (hunk 43 above)
Comment (skorodenko) [reply hangman.py:339]: Constants removed.

File: hangman.py [hunk 49]
```diff
@@ -0,0 +1,400 @@
...
+class BaseVariables():
+    """
+    This class creates an object with base variables.
+    It also has constants inside of it.
+    """
+    vowels = VOWELS_LIST
+    hint = HINT_SYMBOL
+    gap = GAP_SYMBOL
+    yes = YES_ANSWER
+    no = NO_ANSWER
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:31]: Is this a critical mistake? If it is not I would like to leave it as it is.

File: hangman.py [hunk 50]
```diff
@@ -0,0 +1,400 @@
...
+        corresponding letters of other_word, or the letter is the special symbol
+        _ , and my_word and other_word are of the same length;
+        False otherwise: 
+    """
+    if len(my_word) == len(other_word):
+        for i in range(len(my_word)):
+            if my_word[i] == BaseVariables.gap.strip():
+                if other_word[i] in letters_guessed:
+                    return False
+                continue
```
[the remark this answers is not in the document]
Comment (skorodenko) [reply hangman.py:205]: Changed.

File: hangman.py [hunk 51]
```diff
@@ -0,0 +1,384 @@
...
+
+class BaseVariables():
+    """
+    This class creates an object with base variables.
+    It also has constants inside of it.
+    """
+    vowels = VOWELS_LIST
+    hint = HINT_SYMBOL
+    gap = GAP_SYMBOL
+    space = GAP_SPACE
```
Comment (Hozhyi) [inline hangman.py:29]: That's a duplication of code. If you really want to have it in this class - just declare these constants in scope of this class

File: hangman.py [hunk 52]
```diff
@@ -0,0 +1,384 @@
...
+import string
+from enum import Enum
+
+WORDLIST_FILENAME = "words.txt"
+VOWELS_LIST = {"a","e","i","o","u"}
+INITIAL_GUESSES = 6
+ININITAL_WARNINGS = 3
+INITIAL_HINT_ATTEMPTS = 1
+GAP_SYMBOL = "_"
+GAP_SPACE = " "
```
Comment (Hozhyi) [inline hangman.py:17]: TBH - I don't think this constant (GAP_SPACE) is a valuable. All other constants - directly affecting the game. But space - is just a space, do you really think you can change in future this SPACE to something differ?

File: hangman.py [hunk 53]
```diff
@@ -0,0 +1,384 @@
...
+    """
+    wordlist (list): list of words (strings)
+    
+    Returns a word from wordlist at random
+    """
+    return random.choice(wordlist)
+
+
+wordlist = load_words()
+BaseVariables.word_list = wordlist
```
Comment (Hozhyi) [inline hangman.py:87]: Effectively - you can just combine these lines

File: hangman.py [hunk 54]
```diff
@@ -0,0 +1,384 @@
...
+    gap: string, how unguessed letters shold look like
+    returns: string, comprised of letters, underscores (_), and spaces that represents
+      which letters in secret_word have been guessed so far.
+    """
+    # getting set of all letters that were not guessed
+    check = set(secret_word).difference(letters_guessed)
+    new_word = []
+    for letter in secret_word:
+        if letter in check:
+            new_word.append(BaseVariables.gap+BaseVariables.space)
```
Comment (Hozhyi) [inline hangman.py:119]: How about having a space before and after `+`?

**Review by Hozhyi — APPROVED** [review]

**Pull request merged**